src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -521,48 +512,52 @@ public int hashCode() {
+        final Object base = NIO_ACCESS.getBufferBase(b);
+        return (base == null)
+                ? arrayFreeSegment(b, offset, length)
+                : NIO_ACCESS.arrayBackedSegment(b, base, offset, length, b.isReadOnly(), bufferScope(b));
+    }
+
+    @ForceInline
+    private static long offset(Buffer b, int scaleFactor) {
+        final long bbAddress = NIO_ACCESS.getBufferAddress(b);
+        return bbAddress + (((long) b.position()) << scaleFactor);
+    }
+
+    @ForceInline
+    private static long length(Buffer b, int scaleFactor) {
+        return ((long) b.limit() - b.position()) << scaleFactor;
+    }
+
+    @ForceInline
+    private static AbstractMemorySegmentImpl arrayFreeSegment(Buffer b, long offset, long length) {
+        if (!b.isDirect()) {
             throw new IllegalArgumentException("The provided heap buffer is not backed by an array.");
         }
-        long bbAddress = NIO_ACCESS.getBufferAddress(bb);
-        UnmapperProxy unmapper = NIO_ACCESS.unmapper(bb);
-
-        int pos = bb.position();
-        int limit = bb.limit();
-        int size = limit - pos;
-
-        AbstractMemorySegmentImpl bufferSegment = (AbstractMemorySegmentImpl) NIO_ACCESS.bufferSegment(bb);
-        boolean readOnly = bb.isReadOnly();
-        int scaleFactor = getScaleFactor(bb);
-        final MemorySessionImpl bufferScope;
-        if (bufferSegment != null) {
-            bufferScope = bufferSegment.scope;
-        } else {
-            bufferScope = MemorySessionImpl.createHeap(bufferRef(bb));
-        }
-        long off = bbAddress + ((long)pos << scaleFactor);
-        long len = (long)size << scaleFactor;
-        if (base != null) {
-            return switch (base) {
-                case byte[]   _ -> new HeapMemorySegmentImpl.OfByte(off, base, len, readOnly, bufferScope);
-                case short[]  _ -> new HeapMemorySegmentImpl.OfShort(off, base, len, readOnly, bufferScope);
-                case char[]   _ -> new HeapMemorySegmentImpl.OfChar(off, base, len, readOnly, bufferScope);
-                case int[]    _ -> new HeapMemorySegmentImpl.OfInt(off, base, len, readOnly, bufferScope);
-                case float[]  _ -> new HeapMemorySegmentImpl.OfFloat(off, base, len, readOnly, bufferScope);
-                case long[]   _ -> new HeapMemorySegmentImpl.OfLong(off, base, len, readOnly, bufferScope);
-                case double[] _ -> new HeapMemorySegmentImpl.OfDouble(off, base, len, readOnly, bufferScope);
-                default         -> throw new AssertionError("Cannot get here");
-            };
-        } else if (unmapper == null) {
-            return new NativeMemorySegmentImpl(off, len, readOnly, bufferScope);
-        } else {
-            return new MappedMemorySegmentImpl(off, unmapper, len, readOnly, bufferScope);
-        }
+        final UnmapperProxy unmapper = NIO_ACCESS.unmapper(b);
+        return unmapper == null
```

Comment:
I would feel more at ease if the creation of segments went through SegmentFactories (this is not a new issue in this PR)

Reply:
https://bugs.openjdk.org/browse/JDK-8343347